File: reference/mark_reserved_indices.py
import torch
from util.save_reference import save_reference
from typing import Tuple, List, Iterable
import einops

def create_tensors(
        shape: Tuple[int, ...] | List[int] | Iterable[int],
        mark = -2,
        dtype: torch.dtype = torch.float32,
        dir: str = "data",
        name: str = "mark_reserved_indices",
        acts = None,
        counts = None,
        trigger = None
):
    x = torch.rand(shape, dtype=dtype) if acts is None else acts
    y = torch.randint(0, 4, shape[1:], dtype=torch.float32) if counts is None else counts
    z = torch.randn(list(shape[:2]) + [1], dtype=dtype) > 0.5 if trigger is None else trigger
    used_values, used_indices = torch.sort(x, dim=-1)
    reservations = used_indices[:,:,-1:]
    reservations[z] = -1
    reservations = einops.rearrange(reservations, 'batch mems flag -> mems (batch flag)')
    sorts, sort_indices = torch.sort(y, dim=-1)
    sorts, sort_indices, reservations
    reservation_mask = reservations != -1
    expanded_usages = sort_indices.unsqueeze(1).expand(-1, reservations.size(1), -1)
    expanded_res = reservations.unsqueeze(-1).expand(-1, -1, y.size(1))
    matches = (expanded_res == expanded_usages) & reservation_mask.unsqueeze(-1)
    matches = matches.any(dim=1)
    return torch.where(matches, mark, sort_indices), sort_indices, x, y, z

def create_mark_reserved_indices(
        shape: Tuple[int, ...] | List[int] | Iterable[int],
        mark = -2,
        dtype: torch.dtype = torch.float32,
        dir: str = "data",
        name: str = "mark_reserved_indices",
):
    avail, all, normal, counts, trigger = create_tensors(shape, mark, dtype, dir, name)
    save_reference(avail, dir, f"{name}_avail")
    save_reference(all, dir, f"{name}_all")
    save_reference(normal, dir, f"{name}_normal")
    save_reference(counts, dir, f"{name}_counts")
    save_reference(trigger, dir, f"{name}_trigger")

if __name__ == "__main__":
    # batch size, nodes, mems
    create_mark_reserved_indices((6, 7, 8), mark=-2, dtype=torch.float32, dir="data", name="mark_reserved_indices")

File: src/functions/mark_reserved_indices.rs
use ndarray::Axis;
use ndarray::Zip;
use std::fmt::Debug;

use super::abs::{self, abs_ndarray};
use super::expand::expand_at_dim;
use super::rearrange::rearrange_batch_mems_flag;
use super::slicing::slice_last_dim;
use super::sort::argsort_last_dim;
use super::sort::sort_last_dim;
use super::unsqueeze::{self, unsqueeze};
use super::wheres::where_op;
use super::wheres::where_value;
use ndarray::{Array, ArrayD, Dimension, IxDyn, RemoveAxis, ScalarOperand};

pub fn mark_reserved_indices<T>(
    acts: &ArrayD<T>,
    usages: &ArrayD<T>,
    trigger_growth: &ArrayD<bool>,
    mark: i64, // should use negative values such as -2 as otherwise may be mistaken for normal usage value
) -> (ArrayD<i64>, ArrayD<i64>)
where
    T: num_traits::Float + Clone + Debug + ScalarOperand,
{
    // let used_values = sort_last_dim(acts);
    let used_indices = argsort_last_dim(acts);
    let reservations = slice_last_dim(&used_indices).unwrap();
    let reservations = where_value(trigger_growth, &reservations, -1).unwrap();
    let reservations = rearrange_batch_mems_flag(&reservations).unwrap();

    // let sorts = sort_last_dim(&usages);
    let sort_indices: ndarray::ArrayBase<
        ndarray::OwnedRepr<i64>,
        ndarray::Dim<ndarray::IxDynImpl>,
    > = argsort_last_dim(&usages);

    let reservation_mask = &reservations.mapv(|x| x != -1);

    let expanded_usages = unsqueeze(&sort_indices, 1);

    let expanded_usages: ndarray::ArrayBase<
        ndarray::OwnedRepr<i64>,
        ndarray::Dim<ndarray::IxDynImpl>,
    > = expand_at_dim(&expanded_usages, 1, reservations.shape()[1]).unwrap();

    let expanded_res = unsqueeze(&reservations, reservations.ndim());

    let expanded_res: ndarray::ArrayBase<
        ndarray::OwnedRepr<i64>,
        ndarray::Dim<ndarray::IxDynImpl>,
    > = expand_at_dim(&expanded_res, 2, usages.shape()[1]).unwrap();

    let unsqueezed_mask: ndarray::ArrayBase<
        ndarray::OwnedRepr<bool>,
        ndarray::Dim<ndarray::IxDynImpl>,
    > = unsqueeze(&reservation_mask, reservation_mask.ndim());

    let comparison = Zip::from(&expanded_res)
        .and(&expanded_usages)
        .map_collect(|res_val_ref, usage_val_ref| res_val_ref == usage_val_ref);

    let matches = Array::from_shape_fn(comparison.raw_dim(), |idx| {
        // Access the mask value at (i, j, 0) - broadcasting the last dimension
        let mask_idx = [idx[0], idx[1], 0];
        // Perform logical AND between comparison at full coordinate and mask with broadcast
        comparison[idx] && unsqueezed_mask[mask_idx]
    });

    let reduced_matches = matches.map_axis(Axis(1), |view| view.iter().any(|&x| x));

    let avail: ndarray::ArrayBase<ndarray::OwnedRepr<i64>, ndarray::Dim<ndarray::IxDynImpl>> =
        where_value(&reduced_matches, &sort_indices, mark).unwrap();

    (avail, sort_indices)
}
